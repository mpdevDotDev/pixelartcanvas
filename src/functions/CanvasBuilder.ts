import {Subject} from 'rxjs';

export type color = string;
export interface CanvasBuilderProps {
  canvasElement: HTMLCanvasElement;
  pixelSize: number;
};

class CanvasBuilder {
  #canvasElement: HTMLCanvasElement;
  #ctx: CanvasRenderingContext2D;
  #pixelSize: number;
  #width: number;
  #height: number;
  #horizontalPixels: number;
  #verticalPixels: number;
  #inputDataObserver: Subject<color[][]>;
  #inputData: color[][];

  get canvasElement(): HTMLCanvasElement {
    return this.#canvasElement;
  }

  get ctx():CanvasRenderingContext2D {
    return this.#ctx;
  }

  get pixelSize(): number {
    return this.#pixelSize;
  }

  get widht(): number {
    return this.#width;
  }

  get height(): number {
    return this.#height;
  }

  get horizontalPixels(): number {
    return this.#horizontalPixels;
  }

  get verticalPixels(): number {
    return this.#verticalPixels;
  }

  set inputData(inputData:color[][]) {
    this.#inputDataObserver.next(inputData);
  };

  constructor({canvasElement, pixelSize}: CanvasBuilderProps) {
    this.#canvasElement = canvasElement;
    this.#ctx = this.#canvasElement.getContext('2d');
    this.#pixelSize = pixelSize;
    this.#inputDataObserver = new Subject<color[][]>();

    this.#determineProprerties();

    this.#inputDataObserver.subscribe({
      next: (inputData) => {
        this.#inputData = inputData;
        this.update();
      },
    });
  }

  #determineProprerties() {
    let widthAtt = this.#canvasElement.getAttribute('width');
    let heightAtt = this.#canvasElement.getAttribute('height');
    widthAtt = widthAtt.replace(/[^0-9]/gi, '');
    heightAtt = heightAtt.replace(/[^0-9]/gi, '');

    this.#width = parseFloat(widthAtt);
    this.#height = parseFloat(heightAtt);

    this.#horizontalPixels = this.#width / this.#pixelSize;
    this.#verticalPixels = this.#height / this.#pixelSize;
  }

  #getColor(inputValue: color) {
    return `#${inputValue.padEnd(3, '0')}`;
  }

  update() {
    let x = 0;
    let y = 0;
    for (let l = 0; l < this.#verticalPixels; l++) {
      y = l * this.#pixelSize;

      for (let r = 0; r < this.#horizontalPixels; r++) {
        const color = this.#getColor(this.#inputData[l][r]);
        x = r * this.#pixelSize;
        this.#ctx.fillStyle = color;
        this.#ctx.fillRect(x, y, this.#pixelSize, this.#pixelSize);
      }
    }
  }
};

export default CanvasBuilder;
