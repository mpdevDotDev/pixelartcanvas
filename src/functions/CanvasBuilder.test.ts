import CanvasBuilder from './CanvasBuilder';
import 'jest-canvas-mock';

describe('CanvasBuilder', () => {
  const canvasWidth = 200;
  const canvasHeight = 200;
  const pixelSize = 40;
  const dataInput = [
    ['000', '888', 'aaa', 'ccc', 'eee'],
    ['000', '880000', 'aa0000', 'cc0000', 'ee0000'],
    ['000', '008800', '00aa00', '00cc00', '00ee00'],
    ['000', '000088', '0000aa', '0000cc', '0000ee'],
    ['000', '000', '000', '000', 'cc88cc'],
  ];

  const canvasElement = document.createElement('canvas');
  const ctx = canvasElement.getContext('2d');
  canvasElement.setAttribute('width', `${canvasWidth}px`);
  canvasElement.setAttribute('height', `${canvasHeight}px`);

  const canvasBuilder = new CanvasBuilder({canvasElement, pixelSize});

  const update = jest.spyOn(canvasBuilder, 'update');

  it('should store the received canvas element and pixelSize', () => {
    expect(canvasBuilder.canvasElement).toEqual(canvasElement);
    expect(canvasBuilder.pixelSize).toEqual(pixelSize);
  });

  it('shold determine width, height, horizontal and vertical pixels', () => {
    expect(canvasBuilder.widht).toEqual(canvasWidth);
    expect(canvasBuilder.height).toEqual(canvasHeight);
    expect(canvasBuilder.horizontalPixels).toEqual(canvasWidth / pixelSize);
    expect(canvasBuilder.verticalPixels).toEqual(canvasHeight / pixelSize);
  });

  it('should get canvas 2d context', () => {
    expect(canvasBuilder.ctx).toBeTruthy;
  });

  it('should call \'update\' when inputData is changed', () => {
    canvasBuilder.inputData = dataInput;
    expect(update).toBeCalled();
  });

  it('should match ctx snapshot', () => {
    const ctxDrawCalls = ctx.__getDrawCalls();
    expect(ctxDrawCalls).toMatchSnapshot();
  });

  it('should match ctx.fillStyle with last dataInput color', () => {
    const lastLine = dataInput[dataInput.length -1];
    const lastPixel = lastLine[lastLine.length -1];
    expect(ctx.fillStyle).toEqual(`#${lastPixel.padEnd(6, '0')}`);
  });
});
