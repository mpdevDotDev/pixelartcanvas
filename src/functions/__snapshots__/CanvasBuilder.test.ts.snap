// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasBuilder should match ctx snapshot 1`] = `
Array [
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 40,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 80,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 120,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 160,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 0,
      "y": 40,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 40,
      "y": 40,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 80,
      "y": 40,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 120,
      "y": 40,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 160,
      "y": 40,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 0,
      "y": 80,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 40,
      "y": 80,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 80,
      "y": 80,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 120,
      "y": 80,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 160,
      "y": 80,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 0,
      "y": 120,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 40,
      "y": 120,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 80,
      "y": 120,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 120,
      "y": 120,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 160,
      "y": 120,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 0,
      "y": 160,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 40,
      "y": 160,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 80,
      "y": 160,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 120,
      "y": 160,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 40,
      "width": 40,
      "x": 160,
      "y": 160,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;
